luarules/gadgets/unit_custom_weapons_overpen.lua

```diff
@@ -0,0 +1,449 @@
+    end
+    local damageBase = min(damage, damageType)
+
+    if weaponData.falloff then
+        damage = damage * damageLeft
+        damageBase = damageBase * damageLeft
+        damageLeft = damageLeft - health / damageBase - weaponData.penalty
+
+        if weaponData.expDefID and damageLeft <= explodeThreshold then
+            explodeDriver(projID, weaponData.expDefID, attackID, unitID, featureID)
+            return damage
+        end
+    end
+
+    if damageBase > health and damageBase >= healthMax * damageThreshold then
+        if damageLeft > 0 then
+            if weaponData.falloff then
+                driver[2] = damageLeft
+            end
+            respawn[projID] = driver
+        end
+        return damage
+    elseif weaponData.expDefID then
+        explodeDriver(projID, weaponData.expDefID, attackID, unitID, featureID)
+        return damage
+    end
+
+    local damageDone = min(1, damageBase / damageType)
+    return damage, inertia(damageDone) * damageDone
+end
+
+---Simulate the overpen effect by creating a new projectile.
+local function spawnPenetrator(projID, attackID, penDefID, unitID, featureID)
+    local penetrator = respawn[projID]
+    respawn[projID] = nil
+
+    local px, py, pz = spGetProjectilePosition(projID)
+    local timeToLive = spGetProjectileTimeToLive(projID)
+    local vx, vy, vz, vw = spGetProjectileVelocity(projID)
+
+    local driverData = penetrator[1]
+    local explodeID = driverData.expDefID
+
+    if driverData.slowing ~= nil then
+        local speedLeft = inertia(penetrator[2])
+        vx, vy, vz, vw = vx * speedLeft,
+                         vy * speedLeft,
+                         vz * speedLeft,
+                         vw * speedLeft
+    end
+
+    if driverData.penDefID ~= nil then
+        -- Spawn an alternative weapondef:
+        penDefID = driverData.penDefID
+        timeToLive = timeToLive * driverData.ttlRatio
+        local velRatio = driverData.velRatio
+        vx, vy, vz, vw = vx * velRatio,
+                         vy * velRatio,
+                         vz * velRatio,
+                         vw * velRatio
+        -- Penetrators may or may not be drivers, themselves:
+        penetrator[1] = weaponParams[penDefID] -- nil if not
+    end
+
+    local mx, my, mz, radius
+    if unitID ~= nil then
+        local _
+        _, _, _, mx, my, mz = spGetUnitPosition(unitID, true)
+        radius = spGetUnitRadius(unitID)
+    elseif featureID ~= nil then
+        local _
+        _, _, _, mx, my, mz = Spring.GetFeaturePosition(featureID, true)
+        radius = Spring.GetFeatureRadius(featureID)
+    end
+
+    -- Get the time to travel to a position opposite the target's sphere collider.
+    local frames = (radius / vw) * (cos(atan((mx - px) / radius - vx / vw)) +
+                                    cos(atan((my - py) / radius - vy / vw)) +
+                                    cos(atan((mz - pz) / radius - vz / vw))) * 2/3
```

Comment:
The `local` keyword inside the if statements is creating a new variable scope. These `mx`, `my`, and `mz` are distinct from the ones declared on line 313. Thus, their values on lines 325-327 will always be nil.

Reply:
Lua parses the two lines separately- so `local _` should produce a LOADNIL, then the line below will MOVE the first three results into that nil's register position.

I just wanted to throw out those values. Lua makes it a little cumbersome to do since `select` is unusually slow

Reply:
Those three*